### PR TITLE
ICMSLST-2745 - Complete DFL Licence PDF document

### PR DESCRIPTION
**Pull request by chris-pettinga**

### CODE REVIEW

File: web/templates/pdf/import/fa-dfl-licence.html [hunk 1]
```diff
@@ -1,22 +1,236 @@
...
+          <ol>
+            <li>If 10 consignments have been imported but the licence is not exhausted, the importer or his agent should
+              return it to Import Licensing Branch with a request for issue of a new licence for the outstanding
+              balance.
+            </li>
+            <li class="mt-1">When the licence is no longer required (whether it has expired or not), or if it is
+              revoked, then
+              return
+              it to<br/>
+              Department for International Trade<br/>
```
Comment (mikeylockwood) [inline web/templates/pdf/import/fa-dfl-licence.html:225]: Department for Business and Trade

File: web/templates/pdf/import/fa-dfl-licence.html [hunk 2]
```diff
@@ -1,22 +1,236 @@
...
+              Queensway House, West Precinct<br/>
+              Billingham, United Kingdom TS23 2NF<br/>{# /PS-IGNORE #}
+              Email: enquiries.ilb@trade.gov.uk{# /PS-IGNORE #}
```
Comment (mikeylockwood) [inline web/templates/pdf/import/fa-dfl-licence.html:228]: Probably pass this into context from settings so we don't have to update it in the certificate when the email changes

File: web/templates/pdf/import/fa-dfl-licence.html [hunk 3]
```diff
@@ -1,22 +1,236 @@
...
+        may
+        be necessary to effect importation will be granted.
+      </p>
+      <br/>
+      <div class="avoid-break">
+        <p class="content mb-1">This licence</p>
+        <ul id="this-licence-bullets">
+          <li>is not transferable and may be used only for importations by the holder named above.</li>
+          <li>may be revoked at any time without reason given.</li>
+          <li>may be altered only by the Department for International Trade.</li>
```
Comment (mikeylockwood) [inline web/templates/pdf/import/fa-dfl-licence.html:145]: Business and Trade

File: web/templates/pdf/import/fa-dfl-licence.html (hunk 2 above)
Comment (mikeylockwood) [reply web/templates/pdf/import/fa-dfl-licence.html:228]: `"ilb_contact_email": settings.ILB_CONTACT_EMAIL` is already passed into `get_sanctions_context`. Just move this into `get_licence_context`

File: web/utils/pdf/pages.py [hunk 4]
```diff
@@ -2,6 +2,7 @@
 from typing import Any
 
 import pypdf
+from PyPDF2 import PageObject
```
Comment (mikeylockwood) [inline web/utils/pdf/pages.py:5]: Use`pypdf.PageObject` or

```
if TYPE_CHECKING:
    from pypdf import PageObject
```

File: web/utils/pdf/pages.py [hunk 5]
```diff
@@ -143,3 +144,65 @@ def format_wood_pages(pdf_data: bytes) -> bytes:
...
+            if page_number >= 1:
+                # add the top closing line to this page if it's not the first.
+                # the first page already has the top line in CSS
+                page = add_top_line(page)
+                writer.add_page(page)
+                should_add_top_line = False
+                continue
+            else:
+                # don't bother with any borders if the entire licence is on 1 page
+                return pdf_data
```
Comment (mikeylockwood) [inline web/utils/pdf/pages.py:193]: If you're not modifying single page pdfs, return after line 154
```
if page_total == 1:
    return pdf_data
```

File: web/utils/pdf/pages.py [hunk 6]
```diff
@@ -143,3 +144,65 @@ def format_wood_pages(pdf_data: bytes) -> bytes:
...
+        new_page.line(left_margin, 5, right_margin, 5)
+        new_page.save()
+        packet.seek(0)
+        new_pdf = pypdf.PdfReader(packet)
+        page.merge_page(new_pdf.pages[0])
+        return page
+
+    should_add_top_line = True
+
+    for page_number, page in enumerate(reader.pages):
```
Comment (mikeylockwood) [inline web/utils/pdf/pages.py:181]: This is all a bit confusing, if you add an early return before the for loop then there will always be a need to add a line on the page. Just draw a canvas once per page and add what lines you need to it

```
for page_number, page in enumerate(reader.pages, start=1):
    packet = io.BytesIO()
    new_page = canvas.Canvas(packet, pagesize=portrait(A4))

    if page_number > 1:
        new_page.line(left_margin, 5, right_margin, 5)

    if page_number < page_total:
        new_page.line(left_margin, 5, right_margin, 5)

    new_page.save()
    packet.seek(0)
    new_pdf = pypdf.PdfReader(packet)
    page.merge_page(new_pdf.pages[0])
```

File: web/utils/pdf/pages.py (hunk 4 above)
Comment (mikeylockwood) [reply web/utils/pdf/pages.py:5]: I would rewrite the for loop that uses the functions that use these types and remove them altogether. See suggestion below

File: web/utils/pdf/pages.py (hunk 6 above)
Comment (mikeylockwood) [reply web/utils/pdf/pages.py:181]: After merging the pages, just add
```
 if "This licence will cease to be valid for importations" in page.extract_text():
     break
```

Comment (mikeylockwood) [reply web/utils/pdf/pages.py:181]: I see, you won't want a bottom line on that page either

```
signature_page = "This licence will cease to be valid for importations" in page.extract_text()
...
if not signature_page and page_number < page_total:
    new_page.line(left_margin, 5, right_margin, 5)
...
if signature_page:
     break
```

Comment (chris-pettinga) [reply web/utils/pdf/pages.py:181]: @mikeylockwood but sometimes there's a nice little disclaimer after the signature page which we want to include in the new PDF but without lines

Comment (chris-pettinga) [reply web/utils/pdf/pages.py:181]: so we need to continue the for loop to get all the pages in

**Review by mikeylockwood — APPROVED** [review]

**Pull request merged**